File: src/Systems/DataLoader.cs
using Godot;
using Newtonsoft.Json;
using System.Collections.Generic;
using System.IO;

namespace Systems
{
    public static class DataLoader
    {
        public static List<ElectionDefinition> LoadElectionDefinitions(string path)
        {
            string fullPath = ProjectSettings.GlobalizePath(path);
            if (!File.Exists(fullPath))
            {
                GD.PrintErr($"Election definitions file not found: {path}");
                return new List<ElectionDefinition>();
            }

            string json = File.ReadAllText(fullPath);
            var data = JsonConvert.DeserializeObject<ElectionDefinitionsData>(json);
            return data.Elections;
        }

        public static List<CandidateDefinition> LoadCandidateDefinitions(string path)
        {
            string fullPath = ProjectSettings.GlobalizePath(path);
            if (!File.Exists(fullPath))
            {
                GD.PrintErr($"Candidate definitions file not found: {path}");
                return new List<CandidateDefinition>();
            }

            string json = File.ReadAllText(fullPath);
            var data = JsonConvert.DeserializeObject<CandidateDefinitionsData>(json);
            return data.Candidates;
        }
    }

    // Helper classes for JSON deserialization
    public class ElectionDefinitionsData
    {
        [JsonProperty("elections")]
        public List<ElectionDefinition> Elections { get; set; }
    }

    public class ElectionDefinition
    {
        [JsonProperty("title")]
        public string Title { get; set; }

        [JsonProperty("type")]
        public string Type { get; set; }

        [JsonProperty("frequency_in_years")]
        public int FrequencyInYears { get; set; }

        [JsonProperty("applicable_regions")]
        public List<string> ApplicableRegions { get; set; }
    }

    public class CandidateDefinitionsData
    {
        [JsonProperty("candidates")]
        public List<CandidateDefinition> Candidates { get; set; }
    }

    public class CandidateDefinition
    {
        [JsonProperty("name")]
        public string Name { get; set; }

        [JsonProperty("election_type")]
        public string ElectionType { get; set; }

        [JsonProperty("platform")]
        public Dictionary<string, float> Platform { get; set; }
    }
}

File: src/Systems/DataLoader.cpp
#include "DataLoader.hpp"
#include <JSONParseResult.hpp>
#include <Array.hpp>
#include <Dictionary.hpp>

namespace Systems {

using namespace godot;

void DataLoader::_register_methods() {
    register_method("load_election_definitions", &DataLoader::load_election_definitions);
    register_method("load_candidate_definitions", &DataLoader::load_candidate_definitions);
}

DataLoader::DataLoader() {}
DataLoader::~DataLoader() {}

void DataLoader::_init() {}

godot::Error DataLoader::load_json_file(const godot::String& path, godot::String& json_content) {
    godot::String full_path = ProjectSettings::get_singleton()->globalize_path(path);
    
    Ref<File> file;
    file.instance();
    
    Error err = file->open(full_path, File::READ);
    if (err != Error::OK) {
        Godot::print_error("File not found: " + path, "load_json_file", __FILE__, __LINE__);
        return err;
    }

    json_content = file->get_as_text();
    file->close();
    
    return Error::OK;
}

std::vector<Election*> DataLoader::load_election_definitions(const godot::String& path) {
    std::vector<Election*> elections;
    godot::String json_content;
    
    if (load_json_file(path, json_content) != Error::OK) {
        return elections;
    }

    JSON json;
    Ref<JSONParseResult> result = json.parse(json_content);
    
    if (result->get_error() != Error::OK) {
        Godot::print_error("Failed to parse election definitions", 
                          "load_election_definitions",
                          __FILE__, __LINE__);
        return elections;
    }

    Dictionary data = result->get_result();
    Array election_data = data["elections"];

    for (int i = 0; i < election_data.size(); i++) {
        Dictionary election_dict = election_data[i];
        
        Election* election = Election::_new();
        election->set_title(election_dict["title"]);
        election->set_frequency(static_cast<int>(election_dict["frequency_in_years"]));
        
        // Parse applicable regions
        Array regions = election_dict["applicable_regions"];
        std::vector<String> applicable_regions;
        for (int j = 0; j < regions.size(); j++) {
            applicable_regions.push_back(regions[j]);
        }
        
        elections.push_back(election);
    }

    return elections;
}

std::vector<Candidate*> DataLoader::load_candidate_definitions(const godot::String& path) {
    std::vector<Candidate*> candidates;
    godot::String json_content;
    
    if (load_json_file(path, json_content) != Error::OK) {
        return candidates;
    }

    JSON json;
    Ref<JSONParseResult> result = json.parse(json_content);
    
    if (result->get_error() != Error::OK) {
        Godot::print_error("Failed to parse candidate definitions", 
                          "load_candidate_definitions",
                          __FILE__, __LINE__);
        return candidates;
    }

    Dictionary data = result->get_result();
    Array candidate_data = data["candidates"];

    for (int i = 0; i < candidate_data.size(); i++) {
        Dictionary candidate_dict = candidate_data[i];
        
        Candidate* candidate = Candidate::_new();
        candidate->set_name(candidate_dict["name"]);
        
        // Parse platform
        Dictionary platform = candidate_dict["platform"];
        Array keys = platform.keys();
        for (int j = 0; j < keys.size(); j++) {
            String key = keys[j];
            float value = platform[key];
            candidate->add_platform_value(key, value);
        }
        
        candidates.push_back(candidate);
    }

    return candidates;
}

} // namespace Systems 